libs/utils/src/properties.c

```diff
@@ -191,324 +476,423 @@ static void parseLine(const char* line, celix_properties_t *props) {
+        celix_version_t* createdVersion = celix_version_createVersionFromString(entry->value);
+        if (createdVersion != NULL) {
+            return createdVersion;
         }
     }
-    return result;
+    return defaultValue == NULL ? NULL : celix_version_copy(defaultValue);
 }
 
-void celix_properties_setDouble(celix_properties_t *props, const char *key, double val) {
-    char buf[32]; //should be enough to store long long int
-    int writen = snprintf(buf, 32, "%f", val);
-    if (writen <= 31) {
-        celix_properties_set(props, key, buf);
-    } else {
-        fprintf(stderr,"buf to small for value '%f'\n", val);
-    }
+celix_status_t
+celix_properties_setVersion(celix_properties_t* properties, const char* key, const celix_version_t* version) {
+    return celix_properties_createAndSetEntry(properties, key, NULL, NULL, NULL, NULL, celix_version_copy(version));
 }
 
-bool celix_properties_getAsBool(const celix_properties_t *props, const char *key, bool defaultValue) {
-    bool result = defaultValue;
-    const char *val = celix_properties_get(props, key, NULL);
-    if (val != NULL) {
-        char buf[32];
-        snprintf(buf, 32, "%s", val);
-        char *trimmed = celix_utils_trimInPlace(buf);
-        if (strncasecmp("true", trimmed, strlen("true")) == 0) {
-            result = true;
-        } else if (strncasecmp("false", trimmed, strlen("false")) == 0) {
-            result = false;
+void celix_properties_setVersionWithoutCopy(celix_properties_t* properties, const char* key, celix_version_t* version) {
+    celix_properties_createAndSetEntry(properties, key, NULL, NULL, NULL, NULL, version);
+}
+
+int celix_properties_size(const celix_properties_t* properties) {
+    return (int)celix_stringHashMap_size(properties->map);
+}
+
+bool celix_properties_equals(const celix_properties_t* props1, const celix_properties_t* props2) {
+    if (props1 == NULL && props2 == NULL) {
+        return true;
+    }
+    if (props1 == NULL || props2 == NULL) {
+        return false;
+    }
+    if (celix_properties_size(props1) != celix_properties_size(props2)) {
+        return false;
+    }
+    CELIX_PROPERTIES_ITERATE(props1, iter) {
+        celix_properties_entry_t* entry2 = celix_properties_getEntry(props2, iter.key);
+        if (entry2 == NULL || !celix_properties_entryEquals(&iter.entry, entry2)) {
+            return false;
         }
     }
-    return result;
+    return true;
 }
 
-void celix_properties_setBool(celix_properties_t *props, const char *key, bool val) {
-    celix_properties_set(props, key, val ? "true" : "false");
-}
+typedef struct {
+    celix_string_hash_map_iterator_t mapIter;
+    const celix_properties_t* props;
+} celix_properties_iterator_internal_t;
+
+celix_properties_iterator_t celix_properties_begin(const celix_properties_t* properties) {
+    CELIX_BUILD_ASSERT(sizeof(celix_properties_iterator_internal_t) <= sizeof(celix_properties_iterator_t));
```

Comment:
This assertion does not guarantee that `memcpy(iter._data, &internalIter, sizeof(internalIter))` is safe. This does not take the offset of `_data` into account.